### PR TITLE
BUG: Fix segfault due to out of bound pointer in floatstatus check

### PR DESCRIPTION
**Pull request by seberg**
This should fix the CI segfaults tracked in #17542.

### CODE REVIEW

**Comment by seberg** [comment]
I am not quite sure whether the minimum/maximum which also calls this function needs similar treatment. It seems to me that loop is only used for reductions and thus is fine?

**Comment by charris** [comment]
Glad you are checking other uses, there were three in the simd file. Looks like the problematic commit was merged last February.

**Comment by charris** [comment]
Looks like we are more likely to get more recent hardware on actions.

**Comment by seberg** [comment]
Yeah, that would explain why it is suddenly so common.  I think the other things are fine, but most of our other code just uses `dimensions`, etc. (variables which are never modified) here. So if that is enough to avoid compiler reordering, (which it probably is due to the volatile flag?), then I should do the same here as it is simpler. E.g.:
```
npy_clear_floatstatus_barrier((char*)dimensions);
```

File: numpy/core/src/umath/simd.inc.src [hunk 1]
```diff
@@ -3583,7 +3583,7 @@ AVX512F_absolute_@TYPE@(@type@ * op,
         ip += 2*@num_lanes@*stride_ip1;
         num_remaining_elements -= 2*@num_lanes@;
     }
-    npy_clear_floatstatus_barrier((char*)op);
+    npy_clear_floatstatus_barrier((char*)(op - @num_lanes@));
```
Comment (seberg) [inline numpy/core/src/umath/simd.inc.src:3586]:
```suggestion
    npy_clear_floatstatus_barrier((char*)&num_remaining_elements);
```
Seems simpler and also correct to me.

**Comment by charris** [comment]
Does the actual value matter as long as it is valid memory? I vaguely recall the discussion about these barrier functions but have forgotten how they worked.

**Comment by seberg** [comment]
The value doesn't matter, the whole point is that it is types as `volatile` so that the process can't reorder the instructions/math. I doubt reordering is likely here anyway, since there is a loop, but I am not sure what reorder magic compilers are capable of.

**Comment by seberg** [comment]
But considering that, I am honestly not sure that a stack allocated value like this is a valid "barrier"...

**Comment by charris** [comment]
ISTR that the only reason the variable was passed in was to make `volatile` work. If so, then it has no other effect.

**Comment by charris** [comment]
Thanks Sebastian, lets see if this fixes things.

**Comment by mattip** [comment]
FWIW, when I put this in I used the [compiler explorer](https://gcc.godbolt.org/) to test out what assembly was emitted by different constructs. I added the local variable reference to force the compiler not to reorder the barrier call. It didn't seem to matter what variable I used, as long as the compiler did not optimize it away.

**Comment by tjb900** [comment]
Thanks very much for fixing this - it's actually been causing us a reasonable amount of grief. Given the simple and low-risk nature of the final change, just wondering if it's possible to reconsider it for inclusion in 1.19.5?

We've worked around most occurrences of this issue with some python-level code changes (e.g. using `np.real(np.conj(x) * x)`) but of course some uses of complex `np.abs` exist in other libraries which are still tripping us up.

Either way, thanks again!

**Comment by seberg** [comment]
@tjb900 it is already included in the 1.19.x branch, just waiting for the actual release.

**Comment by tjb900** [comment]
Ah, of course - I misinterpreted the label changes.  Thanks.

**Pull request merged**